File: examples/python/batch.py
#!/usr/bin/env python3
"""
Batch reverse-geocode up to 100 points in one HTTP call.

Why batch?
----------
Backfills and imports often have thousands of coordinates. Batching reduces
HTTP overhead while billing stays simple: 1 unit per point (same as single reverse).

Run:
    export REVADDR_API_KEY="sk_live_..."
    python batch.py
"""

from __future__ import annotations

import json
import os
import sys
from typing import Any

import requests

DEFAULT_BASE = "https://api.revaddr.com"

# Sample CONUS points (DC, San Francisco, NYC).
SAMPLE_POINTS = [
    {"lat": 38.8977, "lon": -77.0365},
    {"lat": 37.7749, "lon": -122.4194},
    {"lat": 40.7128, "lon": -74.0060},
]


def reverse_batch(
    points: list[dict[str, float]],
    *,
    api_key: str,
    base_url: str,
) -> list[dict[str, Any]]:
    """
    POST /v1/reverse/batch with {"points": [{"lat": ..., "lon": ...}, ...]}.

    Returns the "results" array (same fields as a single reverse "result").
    """
    if not points:
        raise ValueError("points must not be empty")
    if len(points) > 100:
        raise ValueError("API allows at most 100 points per batch request")

    url = f"{base_url.rstrip('/')}/v1/reverse/batch"
    headers = {
        "x-api-key": api_key,
        "Content-Type": "application/json",
        "Accept": "application/json",
    }
    response = requests.post(
        url,
        headers=headers,
        json={"points": points},
        timeout=(5, 60),
    )
    response.raise_for_status()
    payload = response.json()
    return payload["results"]


def main() -> int:
    api_key = os.environ.get("REVADDR_API_KEY", "").strip()
    if not api_key:
        print("Set REVADDR_API_KEY.", file=sys.stderr)
        return 1

    base_url = os.environ.get("REVADDR_BASE_URL", DEFAULT_BASE).strip() or DEFAULT_BASE

    try:
        results = reverse_batch(SAMPLE_POINTS, api_key=api_key, base_url=base_url)
    except requests.HTTPError as exc:
        body = ""
        if exc.response is not None:
            body = (exc.response.text or "")[:500]
        print(f"HTTP error: {exc}\n{body}", file=sys.stderr)
        return 1
    except (requests.RequestException, KeyError, TypeError, ValueError) as exc:
        print(f"Request failed: {exc}", file=sys.stderr)
        return 1

    for i, (point, result) in enumerate(zip(SAMPLE_POINTS, results)):
        addr = result.get("formatted_address") or "(none)"
        print(f"[{i}] ({point['lat']}, {point['lon']}) -> {addr}")

    print(json.dumps(results, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/typescript/batch.ts
/**
 * Typed batch reverse for RevAddr (up to 100 points per request).
 *
 * Run:
 *   export REVADDR_API_KEY=sk_live_...
 *   npx tsx batch.ts
 */

import type { ReverseResult } from "./reverse.ts";

const DEFAULT_BASE = "https://api.revaddr.com";

const SAMPLE_POINTS = [
  { lat: 38.8977, lon: -77.0365 },
  { lat: 37.7749, lon: -122.4194 },
  { lat: 40.7128, lon: -74.006 },
];

interface BatchEnvelope {
  results: ReverseResult[];
}

export async function reverseBatch(
  points: Array<{ lat: number; lon: number }>,
  options: { apiKey: string; baseUrl?: string },
): Promise<ReverseResult[]> {
  if (points.length === 0) throw new Error("points must not be empty");
  if (points.length > 100) throw new Error("max 100 points per batch");

  const baseUrl = (options.baseUrl ?? DEFAULT_BASE).replace(/\/$/, "");
  const url = new URL(`${baseUrl}/v1/reverse/batch`);

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 60_000);

  let res: Response;
  try {
    res = await fetch(url, {
      method: "POST",
      headers: {
        "x-api-key": options.apiKey,
        "Content-Type": "application/json",
        Accept: "application/json",
      },
      body: JSON.stringify({ points }),
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timer);
  }

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`HTTP ${res.status}: ${text.slice(0, 500)}`);
  }

  const data = (await res.json()) as BatchEnvelope;
  return data.results;
}

async function main(): Promise<void> {
  const apiKey = (process.env.REVADDR_API_KEY ?? "").trim();
  if (!apiKey) {
    console.error("Set REVADDR_API_KEY.");
    process.exit(1);
  }
  const baseUrl = (process.env.REVADDR_BASE_URL ?? DEFAULT_BASE).trim();

  const results = await reverseBatch(SAMPLE_POINTS, { apiKey, baseUrl });
  results.forEach((r, i) => {
    const p = SAMPLE_POINTS[i];
    console.log(
      `[${i}] (${p.lat}, ${p.lon}) -> ${r.formatted_address ?? "(none)"}`,
    );
  });
  console.log(JSON.stringify(results, null, 2));
}

main().catch((err: unknown) => {
  console.error(err instanceof Error ? err.message : err);
  process.exit(1);
});
